src/misc/nc2xy.1
.TH NC2XY 1 "Feb 27 2014" "GMT 4.5.13 (SVN)" "Generic Mapping Tools"
.SH NAME
nc2xy \- Converting netCDF column file(s) to ASCII xy data
.SH SYNOPSIS
\fBnc2xy\fP \fIfiles\fP [ \fB\-F\fP\fIvar1/var2/...\fP ] [ \fB\-S\fP[\fBr\fP] ] [ \fB\-V\fP ]
[ \fB\-f\fP\fIcolinfo\fP ] [ \fB\-b\fP\fBo\fP ] 
.SH DESCRIPTION
\fBnc2xy\fP reads one or more netCDF files with column data and writes out those columns in ASCII format
to standard output, so that they can be used by \fBpsxy\fP, \fBpsxyz\fP, or \fBxyz2grd\fP.
Modify the precision of the ASCII output format by editing the \fBD_FORMAT\fP parameter in your \.gmtdefaults4 file
or use \fB\-\-D_FORMAT\fP=\fIvalue\fP on the command line.
.TP
\fIfiles\fP
Names of netCDF files to be converted.
.SH OPTIONS
.TP
\fB\-F\fP
Specify up to 10 names of the variables (separated by slashes) to be printed out. All variables to be
1-dimensional and be of equal length. When omited, the first two variables in the netCDF file will be printed.
.TP
\fB\-S\fP
Suppress output for records with one or more NaN values [Default outputs all nodes].
Append \fBr\fP to reverse the suppression, i.e., only output the records with at least one NaN value.
.TP
\fB\-V\fP
Selects verbose mode, which will send progress reports to stderr [Default runs "silently"].
.TP
\fB\-bo\fP
Selects binary output.
Append \fBs\fP for single precision [Default is \fBd\fP (double)].
Uppercase \fBS\fP or \fBD\fP will force byte-swapping.
Optionally, append \fIncol\fP, the number of desired columns in your binary output file.
.TP
\fB\-f\fP
Special formatting of input and/or output columns (time or geographical data).
Specify \fBi\fP or \fBo\fP to make this apply only to input or output [Default applies to both].
Give one or more columns (or column ranges) separated by commas.
Append \fBT\fP (absolute calendar time), \fBt\fP (relative time in chosen \fBTIME_UNIT\fP since \fBTIME_EPOCH\fP),
\fBx\fP (longitude), \fBy\fP (latitude), or \fBf\fP (floating point) to each column
or column range item.  Shorthand \fB\-f\fP[\fBi\fP|\fBo\fP]\fBg\fP means \fB\-f\fP[\fBi\fP|\fBo\fP]0\fBx\fP,1\fBy\fP
(geographic coordinates).
See also \fBTIME COORDINATES\fP below.
.SH TIME COORDINATES
Time coordinates in netCDF files will be recognized as such.
The variable's \fBunit\fP attribute is parsed to determine the unit and epoch of the time coordinate in the grid.
Values are then converted to the internal time system specified by \fBTIME_UNIT\fP and \fBTIME_EPOCH\fP in the \.gmtdefaults
file or on the command line.
The default output is relative time in that time system, or absolute time when using the option \fB\-f0T\fP, \fB\-f1T\fP, etc.
.SH EXAMPLES
To print out latitude, longitude and height in a netCDF file as ASCII records, while suppressing all NaN values:
.br
.sp
\fBnc2xy\fP \fB\-F\fP lat/lon/height \fB\-S\fP trackfile.nc > trackfile.xy
.SH "SEE ALSO"
.IR gmtdefaults (1),
.IR GMT (1),
.IR psxy (1),
.IR psxyz (1),
.IR xyz2grd (1)


ChangeLog
-----------------------------------------------------------------------------------------
	$Id: ChangeLog 12925 2014-02-20 22:02:24Z pwessel $
-----------------------------------------------------------------------------------------

-----------------------------------------------------------------------------------------
201x-yy-zz	pwessel
	GMT 5.1.1 finalized and released.
-----------------------------------------------------------------------------------------
2014-02-20	pwessel
	* grdcontour.c:		The -Zscale/offset option was applied incorrectly due to a
				change in the underlying scaling function.
	* grdvolume.c:		---"---
2014-02-10	pwessel
	*gmt2kml.c:		Make sure the default style IDs are unique for each process.

2014-02-03	pwessel
	* pslib.c:		Auto-widen paragraph width to at least fit widest word.

2014-01-26	pwessel
	* gmt_grdio.c:		Building the grid command string could exceed its length if
				the argument became exactly the max length of the array; then
				a final extra character \0 would be added and exceed array length.
				Thanks to Joachim Saul for pointing it out.

2014-01-22	pwessel
	* pscoast.c:		The processing of -I -N levels and pens could get out of sync
				so that the wrong pen was used for the specified feature.
2014-01-16	remko
	* pslib.c:		Take superscripts and subscripts into account when determining
				dimensions of text box.

2014-01-10	pwessel
	* grdmath.c:		We consider the change in SDIST, PDIST to return degrees instead
				of km to be a bug as it makes it difficult to handle geodesics
				for which degrees make no sense.  Thus we have consolidated all
				the ?DIST operators to give km if geographic data are involved.
				Example 23 has been adjusted accordingly, as has several test
				scripts.  Please check your own scripts for SDIST.

2014-01-09	pwessel
	* gmt_vector.c:		Add protection against introduced longitude-jumps produced
				by resampling in GMT_fix_up_path.
2014-01-09	remko
	* gmt_vector.c:		Add protection against introduced longitude-jumps produced
				by resampling in GMT_fix_up_path.
	* grdmath.c:		LDIST and PDIST will again produce km when using geodesics.

2014-01-01	pwessel
	* x2sys/x2sys.c:	Mishandled the assignment of segment number for each record.
	* gmt_support.c:	Allow GMT_crossover to deal with global wrapped data.

2013-12-31	pwessel
	* gmt2kml.c:		Now using GMT_Put_Record instead of old stdout printing.
				This addresses issue #473.

2013-12-10	pwessel
	* grdvector.c:		Vector direction did not adjust when negative scales were
				used in -JX or -Jx.
	* psxy[z].c		---"---

2013-12-08	remko
	* psimage.c:		Now properly plots transparent images and makes -Gt work.

2013-11-26	pwessel
	* gmt_io.c:		GMT would crash on a blank line with no newline at end at all.
	* pstext.c:		pstext did not process GMT4-style input properly.

2013-11-26	jluis
	* potential/gravfft.c:	Let -E options default to 3.
				Let users select the -N+l (leave trend alone). That is, do not overwrite with default mode.
				Let the default detrend be remove MID point.
				Explicitly set -N+a in gravfft command to avoid failures due to coming changes in gravfft.
2013-11-25	pwessel
	* grdlandmask.c:	Failed to fill partial bins with no coastlines when the region
				crossed Greenwich [issue #451].

2013-11-21	pwessel
	* pscontour.c:		Let pscontour output contours to file [issue #448].
	* pscoast.c:		Prevent users from not giving proper projection [issue #449].

2013-11-19	jluis
	* mgd77/mgd77magref.c:	Needed to reset the t_col variable when constant altitude [issue #446].

2013-11-19	pwessel
	* gmt_sph.c:		Fix invalid read in ssrfpack_grid due to par array being of length 2 rather than 3.

2013-11-18	pwessel
	* gmt_grdio.c:		Place the old auto-detect of 4-byte binary float vs int under compatibility mode for backwardness.

2013-11-16	pwessel
	* grdvector.c:		Address shortcomings of geovectors when wrapped and clipped.

2013-11-15	pwessel
	* geovectors.c:		Fix geovector when shrinking is selected.

2013-11-13	pwessel
	* gmt_api.c:		Fix API issues pointed out by Ian.

2013-11-13	fwobbe
	* GMT.app:		Inherit user shell in GMT.app.

2013-11-12	pwessel
	* gmt_init.c:		grdmath could corrupt the gmt.history file [issue # 438].
 	
2013-11-10	pwessel
	* gmt_support.c:	Fix symbol-size issue in psxy -Sc [issue #435].
	
2013-11-09	pwessel
	* grdvector.c:		Did not convert direction to azimuths for geo vectors.

2013-11-08	pwessel
	* grdvector.c:		Fix grdvector -I problem [issue #434].
	* ps2raster.c:		Under Mavericks, strcpy would exit if the to and from pointers
-----------------------------------------------------------------------------------------
2013-11-05	pwessel
	GMT 5.1.0 finalized and released.
-----------------------------------------------------------------------------------------

2013-11-01	remko
	* example_10.sh:	Changed input to number of languages; multi-color columns

2013-10-31	pwessel
	* gmt_dcw.c:		Added option to specify code as =<continent> which will
				automatically be expanded to all the countries in that
				continent.  E.g., -F=AF+gred will paint Africa read.
				Fixed memory leaks.

2013-10-23	pwessel
	* gmtinfo.c:		The new name of minmax.c, the latter will be honored
				when GMT compatibility mode is 4.
	* grdmath.c:		Fixed porting bugs for CURV and D2DY2; added test ops.sh

2013-10-22	pwessel
	* minmax.c:		Added -D to adjust region found with -I
	* makecpt.c:		Better reporting of gaps or overlaps in CPTs

2013-10-05	pwessel
	* greenspline.c:	Needed -f in order to select -f0T input, plus it made assumptions
				about lon,lat despite not being selected.

2013-10-04	pwessel
	* gmt.history:		This replaces the old .gmtcommands file name.

2013-09-28	remko
	* gmt_init.c:		Moved origin of JD by one day (24 Nov instead of 25 Nov).
	* ConfigUserTemplate.cmake:	Added option GMT_DOCS_DEPEND_ON_GMT

2013-09-27	pwessel
	* share/cpt:		Added 10 new CPTs, mostly for bathymetry and topography.

2013-09-14	pwessel

src/misc/nc2xy.c
/*--------------------------------------------------------------------
 *	$Id: nc2xy.c 10173 2014-01-01 09:52:34Z pwessel $
 *
 *	Copyright (c) 2006-2014 by R. Scharroo
 *	See LICENSE.TXT file for copying and redistribution conditions.
 *
 *	This program is free software; you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation; version 2 or any later version.
 *
 *	This program is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	Contact info: gmt.soest.hawaii.edu
 *--------------------------------------------------------------------*/
/*
 * nc2xy.c reads a netCDF file and prints out the x,y values to
 * standard output.
 *
 * Author:	Remko Scharroo
 * Date:	28-Nov-2005
 * Version:	1
 */

#define NVAR 10
#define GMT_WITH_NO_PS
#include "gmt.h"

int main (int argc, char **argv)
{
	GMT_LONG error = FALSE, suppress = FALSE, reverse = FALSE, skip, plus_const_col = FALSE;
	int ncid, varid[NVAR], dimid = 0, ndims;
	GMT_LONG i, k, n_files = 0, n_out;
	int n_dimid = -1;
	size_t n_total = 0, j, n_suppressed = 0;
	double w, e, s, n, out[NVAR+1], add_offset[NVAR], scale_factor[NVAR], missing_value[NVAR];
	double cte_col = 0.0;
	char varnm[NVAR][GMT_TEXT_LEN], long_name[GMT_LONG_TEXT], units[GMT_LONG_TEXT];
	struct GMT_TIME_SYSTEM time_system;

	argc = (int)GMT_begin (argc, argv);

	w = e = s = n = 0.0;

	for (i = 0; i < NVAR; i++) {
		varnm[i][0] = '\0';
	}

	for (i = 1; i < argc; i++) {
		if (argv[i][0] == '-') {
			switch (argv[i][1]) {
				/* Common parameters */

				case 'f':
				case 'b':
				case 'V':
				case '\0':
					error += GMT_parse_common_options (argv[i], &w, &e, &s, &n);
					break;

				/* Supplemental options */

				case 'C':
					cte_col = atof(&argv[i][2]);
					plus_const_col = TRUE;
					break;
				case 'F':
					sscanf (&argv[i][2], "%[^/]/%[^/]/%[^/]/%[^/]/%[^/]/%[^/]/%[^/]/%[^/]/%[^/]/%[^/]", varnm[0], varnm[1], varnm[2], varnm[3], varnm[4], varnm[5], varnm[6], varnm[7], varnm[8], varnm[9]);
					break;
				case 'S':
					suppress = TRUE;
					if (argv[i][2] == 'r') reverse = TRUE;
					break;
				default:
					error = TRUE;
					GMT_default_error (argv[i][1]);
					break;
			}
		}
		else
			n_files++;
	}

	if (argc == 1 || GMT_give_synopsis_and_exit) {
		fprintf (stderr, "nc2xy %s - Converting netCDF column file(s) to ASCII xy data\n\n", GMT_VERSION);
		fprintf( stderr, "usage: nc2xy <files> [-F<var1>/<var2>/...] [-S[r]] [-V] [%s] [-bo] > xyfile\n", GMT_f_OPT);

		if (GMT_give_synopsis_and_exit) exit (EXIT_FAILURE);

		fprintf (stderr, "\n\t<files> is one or more netCDF files to convert\n");
		fprintf (stderr, "\n\tOPTIONS:\n");
		fprintf (stderr, "\t-F Specify variables to be extracted (up to %d)\n", NVAR);
		fprintf (stderr, "\t-S Suppress records with NaN values [Default prints all nodes]\n");
		fprintf (stderr, "\t   Append r to reverse the suppression (only output records with NaNs)\n");
		GMT_explain_option ('V');
		GMT_explain_option ('f');
		GMT_explain_option ('o');
		GMT_explain_option ('.');
		exit (EXIT_FAILURE);
	}

	if (n_files == 0) {
		fprintf (stderr, "%s: GMT SYNTAX ERROR:  Must specify at least one input file\n", GMT_program);
		error++;
	}

	if (error) exit (EXIT_FAILURE);

#ifdef SET_IO_MODE
	GMT_setmode (GMT_OUT);
#endif

	for (k = 1; k < argc; k++) {
		if (argv[k][0] == '-') continue;	/* Skip the options */

		/* Open netCDF file */

		if (!strcmp (argv[k],"=")) {
			fprintf (stderr, "%s: %s [%s]\n", GMT_program, GMT_strerror(GMT_GRDIO_NC_NO_PIPE), argv[k]);
			exit (EXIT_FAILURE);
		}
		GMT_err_fail (nc_open (argv[k], NC_NOWRITE, &ncid), argv[k]);

		if (gmtdefs.verbose) fprintf (stderr, "%s: Working on file %s\n", GMT_program, argv[k]);

		/* Determine IDs of requested variables; take first two if none given */

		for (n_out = 0; n_out < NVAR && varnm[n_out][0]; n_out++)
			GMT_err_fail (nc_inq_varid (ncid, varnm[n_out], &varid[n_out]), argv[k]);

		if (n_out == 0) {
			n_out = 2;
			varid[0] = 0;
			varid[1] = 1;
		}

		/* Get further variable sizes and attributes */

		for (i = 0; i < n_out; i++) {
			/* Check that variable is 1-dimensional */
			GMT_err_fail (nc_inq_varndims (ncid, varid[i], &ndims), argv[k]);
			if (ndims != 1) {
				fprintf (stderr, "%s: Variable %s is not 1-dimensional\n", GMT_program, varnm[i]);
				exit (EXIT_FAILURE);
			}
			/* Check that dimension is same as the others */
			GMT_err_fail (nc_inq_vardimid (ncid, varid[i], &dimid), argv[k]);
			if (n_dimid < 0) {
				n_dimid = dimid;
				GMT_err_fail (nc_inq_dimlen (ncid, dimid, &n_total), argv[k]);
			}
			else if (n_dimid != dimid) {
				fprintf (stderr, "%s: Variable %s has different dimension than others\n", GMT_program, varnm[i]);
				exit (EXIT_FAILURE);
			}
			/* Check all attributes */
			if (nc_get_att_double (ncid, varid[i], "scale_factor", &scale_factor[i])) scale_factor[i] = 1.0;
			if (nc_get_att_double (ncid, varid[i], "add_offset", &add_offset[i])) add_offset[i] = 0.0;
			if (nc_get_att_double (ncid, varid[i], "_FillValue", &missing_value[i]) &&
			    nc_get_att_double (ncid, varid[i], "missing_value", &missing_value[i])) missing_value[i] = GMT_d_NaN;
			if (GMT_nc_get_att_text (ncid, varid[i], "long_name", long_name, (size_t)GMT_LONG_TEXT)) long_name[0] = 0;
			if (GMT_nc_get_att_text (ncid, varid[i], "units", units, (size_t)GMT_LONG_TEXT)) units[0] = 0;
			GMT_str_tolower (long_name); GMT_str_tolower (units);

			/* Scan for geographical or time units */
			if (!strcmp (long_name, "longitude") || strstr (units, "degrees_e"))
				GMT_io.in_col_type[i] = GMT_IS_LON;
			else if (!strcmp (long_name, "latitude") || strstr (units, "degrees_n"))
				GMT_io.in_col_type[i] = GMT_IS_LAT;
			else if (!strcmp (long_name, "time") || !strcmp (varnm[i], "time")) {
				GMT_io.in_col_type[i] = GMT_IS_RELTIME;
				memcpy ((void *)&time_system, (void *)&gmtdefs.time_system, sizeof (struct GMT_TIME_SYSTEM));
				if (GMT_get_time_system (units, &time_system) || GMT_init_time_system_structure (&time_system))
					fprintf (stderr, "%s: Warning: Time units [%s] in NetCDF file not recognised, defaulting to gmtdefaults.\n", GMT_program, units);
				/* Determine scale between data and internal time system, as well as the offset (in internal units) */
				scale_factor[i] = time_system.scale * gmtdefs.time_system.i_scale;
				add_offset[i] = (time_system.rata_die - gmtdefs.time_system.rata_die) + (time_system.epoch_t0 - gmtdefs.time_system.epoch_t0);
				add_offset[i] *= GMT_DAY2SEC_F * gmtdefs.time_system.i_scale;
			}
			else if (GMT_io.in_col_type[i] == GMT_IS_UNKNOWN)
				GMT_io.in_col_type[i] = GMT_IS_FLOAT;
		}

		/* Load data record by record and scale as required */

		for (j = 0; j < n_total; j++) {
			skip = FALSE;
			for (i = 0; i < n_out; i++) {
				GMT_err_fail (nc_get_var1_double (ncid, varid[i], &j, &out[i]), argv[k]);
				if (out[i] == missing_value[i])
					out[i] = GMT_d_NaN;
				else
					out[i] = out[i] * scale_factor[i] + add_offset[i];
				if (suppress && GMT_is_dnan (out[i])) {
					skip = TRUE;
					continue;
				}
			}
			if (plus_const_col) out[i] = cte_col;	/* Extra column with a const value on output */
			if (skip == reverse)
				GMT_output (GMT_stdout, n_out+plus_const_col, out);
			else
				n_suppressed++;
		}
	}

	if (gmtdefs.verbose) fprintf (stderr, "%s: %ld values extracted\n", GMT_program, (GMT_LONG)(n_total - n_suppressed));
	if (n_suppressed && gmtdefs.verbose) {
		if (reverse)
			fprintf (stderr, "%s: %ld finite values suppressed\n", GMT_program, (GMT_LONG)n_suppressed);
		else
			fprintf (stderr, "%s: %ld NaN values suppressed\n", GMT_program, (GMT_LONG)n_suppressed);
	}

	GMT_end (argc, argv);

	exit (EXIT_SUCCESS);
}
